src/main/java/KilimanJARo/P2P/client/tunneling/Tunnel.java

```diff
@@ -1,28 +1,140 @@
+            e.printStackTrace();
+            return false;
+        }
+    }
+
+    /**
+     * Waits for an incoming connection on the server socket. Once a connection is established,
+     * it checks if the incoming connection's IP matches the expected IP. If it matches, it starts
+     * forwarding data between the sockets.
+     */
+    private void waitForConnection(String to) {
+        for (int i = 0; i < toRetryCount; i++) {
+            try {
+                Socket incomingSocket = serverSocket.accept();
+                if (incomingSocket.getInetAddress().getHostAddress().equals(to.split(":")[0])) {
+                    clientSocket = incomingSocket;
+                    executorService.submit(() -> forwardData(clientSocket, incomingSocket));
+                    executorService.submit(() -> forwardData(incomingSocket, clientSocket));
+                    return;
+                } else {
+                    incomingSocket.close();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        System.err.println("Failed to establish connection after multiple attempts.");
+    }
+
+    /**
+     * Forwards data between the input and output sockets. Reads data from the input socket
+     * and writes it to the output socket.
+     *
+     * @param inputSocket The socket to read data from.
+     * @param outputSocket The socket to write data to.
+     */
+    private void forwardData(Socket inputSocket, Socket outputSocket) {
+        try (var inputStream = inputSocket.getInputStream();
+             var outputStream = outputSocket.getOutputStream()) {
+            byte[] buffer = new byte[4096];
+            int bytesRead;
+            while ((bytesRead = inputStream.read(buffer)) != -1) {
+                outputStream.write(buffer, 0, bytesRead);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
     }
 
     /**
      * Close all connections
      */
-    public void Close() {
-        throw new RuntimeException("Not implemented.");
+    public void close() {
+        try {
+            if (clientSocket != null) {
+                clientSocket.close();
+            }
+            if (serverSocket != null) {
+                serverSocket.close();
+            }
+            executorService.shutdownNow();
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to close tunnel.", e);
+        }
     }
 
     /**
-     * @return Returns port for output connection
+     * @return Returns port for connection to "from"
      */
-    public int Port() {
-        throw new RuntimeException("Not implemented.");
+    public int PortToFrom() {
+        return clientSocket != null ? clientSocket.getLocalPort() : -1;
     }
+
+    /**
+     * @return Returns port for connection to "to"
+     */
+    public int PortToTo() {
+        return serverSocket != null ? serverSocket.getLocalPort() : -1;
```

Comment:
ToTo sounds funny, but i think better to rename =)

Reply:
Changed to PortToPrev and PortToNext